examples/qalora_finetuning/qalora_gptq_finetuning.py

```diff
@@ -0,0 +1,422 @@
+
+    Args:
+        model: The PEFT model with QALoRA adapters
+    """
+    print("\n=== QALoRA Implementation Verification ===")
+
+    total_params = 0
+    trainable_params = 0
+    qalora_modules = 0
+
+    # Check trainable parameters
+    for name, param in model.named_parameters():
+        total_params += param.numel()
+        if param.requires_grad:
+            trainable_params += param.numel()
+
+    # Check for QALoRA modules
+    for name, module in model.named_modules():
+        if hasattr(module, "qalora_group_size"):
+            qalora_modules += 1
+            adapter_name = next(iter(module.qalora_group_size.keys()), None)
+            if adapter_name:
+                print(f"\nQALoRA module found: {name}")
+                print(f"  - Group size: {module.qalora_group_size[adapter_name]}")
+
+                # Check adapter matrices
+                if hasattr(module, "lora_A") and adapter_name in module.lora_A:
+                    a_shape = module.lora_A[adapter_name].weight.shape
+                    a_grad = module.lora_A[adapter_name].weight.requires_grad
+                    print(f"  - lora_A shape: {a_shape}, requires_grad: {a_grad}")
+
+                if hasattr(module, "lora_B") and adapter_name in module.lora_B:
+                    b_shape = module.lora_B[adapter_name].weight.shape
+                    b_grad = module.lora_B[adapter_name].weight.requires_grad
+                    print(f"  - lora_B shape: {b_shape}, requires_grad: {b_grad}")
+
+                # Check base weights
+                if hasattr(module, "weight"):
+                    base_grad = module.weight.requires_grad
+                    print(f"  - Base weight requires_grad: {base_grad} (should be False)")
+
+    # Print summary
+    print("\nParameter summary:")
+    print(f"  - Total parameters: {total_params:,}")
+    print(f"  - Trainable parameters: {trainable_params:,} ({trainable_params / total_params * 100:.2f}%)")
+    print(f"  - QALoRA modules found: {qalora_modules}")
+
+    if qalora_modules > 0 and trainable_params > 0:
+        print("✅ QALoRA implementation verification: PASSED")
+    else:
+        print("❌ QALoRA implementation verification: FAILED")
+
+
+def print_parameter_memory(model, optimizer_name: str = "adam") -> None:
+    """
+    Print detailed memory usage information for model parameters.
+
+    Args:
+        model: The model to analyze
+        optimizer_name: The optimizer type to estimate memory for ("adam" or "sgd")
+    """
+    total_param_mem = 0
+    total_grad_mem = 0
+    total_opt_mem = 0
+
+    print(f"{'Name':50} {'Shape':20} {'Dtype':10} {'Param MB':10} {'Grad MB':10} {'Requires Grad'}")
+    print("-" * 110)
+
+    for name, param in model.named_parameters():
+        param_size = param.numel() * param.element_size() / 1024**2  # in MB
+        grad_size = param_size if param.requires_grad else 0
+        opt_size = 0
```

Comment:
If the user uses neither Adam nor SGD, this would be shown as 0, which would be confusing. Instead, you could check if the value is 0 and then replace it, for instance, with `"unknown"`.